### PR TITLE
bittorrent-dht: add bittorrent-dht app layer v4

### PR DESCRIPTION
**Pull request by abungay**
Parses and logs the bittorrent-dht protocol. 

Feature: #3086

Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/3086

Describe changes:
- Ensures BitTorrent DHT traffic does not get marked as app_proto: failed
- Parses BitTorrent DHT packets and logs data to eve.json

Changes since last PR:
- rebase to master, fix merge conflicts

Reference BitTorrent DHT specification:
https://www.bittorrent.org/beps/bep_0005.html

TODO after merging PR:
- Update wiki for SID ranges at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/AppLayer for the newly introduced bittorrent-dht SID range (start=2271000, end=2271999)

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):

Previous PR: https://github.com/OISF/suricata/pull/5942

suricata-verify-pr: 560

### CODE REVIEW

**Comment by abungay** [comment]
Question for reviewers: It was noticed that the port range "[1024:65535]" being used in rust/src/bittorrent_dht/bittorrent_dht.rs:rs_bittorrent_dht_udp_register_parser() is causing the Suricata startup time to take over 3 minutes on my computer compared to under 1 second on the master Suricata branch. Is there a better way I could listen on ports 1024:65535 without causing such a long startup time? This startup time was causing CI failures for fuzzing on the previous version of this PR v3.

File: rust/src/bittorrent_dht/bittorrent_dht.rs [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+        get_events: Some(rs_bittorrent_dht_state_get_events),
+        get_eventinfo: Some(BitTorrentDHTEvent::get_event_info),
+        get_eventinfo_byid: Some(BitTorrentDHTEvent::get_event_info_by_id),
+        localstorage_new: None,
+        localstorage_free: None,
+        get_files: None,
+        get_tx_iterator: Some(rs_bittorrent_dht_state_get_tx_iterator),
+        get_tx_data: rs_bittorrent_dht_get_tx_data,
+        apply_tx_config: None,
+        flags: 0,
```
Comment (victorjulien) [inline rust/src/bittorrent_dht/bittorrent_dht.rs:345]: transactions look unidirectional, so `APP_LAYER_PARSER_OPT_UNIDIR_TXS` should be set here

**Review by victorjulien — CHANGES_REQUESTED** [review]
Some smaller comments. Another thing that will come up soonish is the nom7 transition.

**Comment by jasonish** [comment]
> Some smaller comments. Another thing that will come up soonish is the nom7 transition.

Actually nom isn't used in this parser.

**Pull request closed without merging**